### ISSUE TITLE
Weights for the constraint model

### ISSUE THREAD

**Issue by fuxuliu** [bytedance/Protenix#75]
Hi, Protenix Team

I noticed that w/ and w/o contact constrain are two models. I have a question, is the weight of the two models exactly the same? Or is it just that the weights are different in the ConstraintEmbedder block, and all the other modules have the same weights?
Then, I would also like to ask, is the w/ constrain model fine-tuned on the basis of the w/o constraint model? Or a brand new one that training from scratch and keeps input with constrain feature?

Looking forward to your reply. Thank you very much.

**Comment by zhangyuxuann** [comment]
@fuxuliu  The w/ constrain model is **fine-tuned**  on the basis of the w/o constraint model. The weight of the two models are different. We haven't tried to train from scratch the constraint model yet. Finetuning is a relatively cheap method to adapt to new features like constraint and esm embedding.

**Comment by fuxuliu** [comment]
@zhangyuxuann Hi, thank you very much for your reply.

I have a few more questions, which may be a little more technical, I hope you don't mind.

You said the w/ constrain model is fine-tuned on the basis of the w/o constraint model, then in the process of fine-tuned, Do you train only one layer of constraint embedder? Or the whole model is unfreezed state?

Another question, I noticed that the contact-constraint feature is actually the contact max distance threshold of the pair of residue-residue (or residue-ligand atom) to be constrained. So how do you do contact-constraint sampling during training, because a bio complex actually has multiple contact interfaces, At the same time, the max distance assigned during training is the real contact distance obtained from the pdb?

**Comment by zhangyuxuann** [comment]
@fuxuliu the whole model is unfreezed state, but the added constraint part and the remaining part(with small learning rate) will have different lr schedule.  @Anfanku can you explain more detail for the another question?

**Comment by fuxuliu** [comment]
@zhangyuxuann Thanks you reply.
@Anfankus Could you please explain more detail for the another question? 
Thank you.

**Comment by Anfankus** [comment]
Hi @fuxuliu, for your question:

> ...how do you do contact-constraint sampling during training

During training, we first sample a `max_distance` and `num_contacts` from an uniform distribution and a geometric distribution respectively. The distributions vary according to the interface type. And then sample `num_contacts` contact pairs within `max_distance` from the ground truth structure.

**Comment by fuxuliu** [comment]
@Anfankus @zhangyuxuann 
okay.
I think I understand a lot. Thank you for your answers

**Comment by fuxuliu** [comment]
@Anfankus Hi, 
I'm sorry to bother again.

And it occurred to me that protein-protein interface (residue-residue), protein-ligand interface(protein-ligand atom), when training, it is generally considered that **distance threshold is** less than how much is it considered that he has contact?
If it is greater than a **distance threshold**, the pairs will not be sampled?

**Comment by Anfankus** [comment]
@fuxuliu 
The distance threshold is less than 30A for protein-protein and is less than 10A for protein-ligand in our default training setting. Residue pairs with spacing greater than the threshold will not be sampled.

**Comment by fuxuliu** [comment]
@zhangyuxuann Hi, Sorry to bother you again.

> the whole model is unfreezed state, but the added constraint part and the remaining part(with small learning rate) will have different lr schedule.

I would like to ask what is the learning rate and the number of steps to fine tune the added constraint part?
If the learning rate is 0.0018, I feel very big? Did the batchsize change from the pretrained phas?

**Comment by zhangyuxuann** [comment]
@fuxuliu The learning rate can be set as 5e-4, we finetune with batch size 64. The steps is about 15-20K.

**Comment by MrBob15** [comment]
@zhangyuxuann Hi, I was wondering whether you used your pretraining data set, or did you use just PDB, as in finetune_demo.sh, and whether it is necessary to include distillation data when fine tuning a constraint model. Thanks!
